### PR TITLE
Follow-up on form-data adjustments for `java`

### PR DESCRIPTION
**Pull request by scbedd**
Contributes to resolving #12259

Timeline of events:

- We merged a PR that treats `multipart/*` the same as `multipart/mixed`.
- Incoming requests that we match against are slow to sanitize, because they are so big.
- We ignore the body anyway in the configured matcher.

So, this PR includes a couple changes.

1. It will detect a multipart boundary from the body content if the sanitized value if `BOUNDARY` now. We didn't used to. (this was the cause of the failure to parse multipart data we were seeing in java openai tests)
2. Now when we are not comparing the body anyway, we do not sanitize the incoming request body based on the activated sanitizers. There is no reason to given that we are ignoring the body during matching anyway.
3. Request/Response bodies will **still** be sanitized when running `Record/Stop`. This is intentional as we don't know if the request body has anything private in it. From a defensive standpoint this is the right thing to do, and it won't reflect negatively on runtime performance.

Will leave this PR open while I publish a version that can be utilized in java CI and confirm that it all checks out.

### CODE REVIEW

**Comment by scbedd** [comment]
[Confirmed fixed!](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=5399889&view=results)

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordMatcher.cs [hunk 1]
```diff
@@ -26,17 +26,24 @@ public class RecordMatcher
...
-        private bool _ignoreQueryOrdering;
+        /// <summary>
+        /// When true, message bodies are compared during matching. In addition, when true incoming request bodies will be sanitized.
+        /// </summary>
+        public bool ShouldCompareBodies { get; set; }
+
+        /// <summary>
+        /// When true, query parameter ordering is ignored during URI normalization.
+        /// </summary>
+        public bool ShouldIgnoreQueryOrdering { get; set; }
```
Comment (alzimmermsft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordMatcher.cs]: Is there a time when we want these to be settable after initialization?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/MultipartUtilities.cs [hunk 2]
```diff
@@ -110,7 +110,7 @@ public static byte[] NormalizeBareLf(byte[] src)
         public static string ResolveFirstBoundary(string boundary, byte[] raw)
         {
             // Boundary might have been sanitised to "REDACTED"
-            if (boundary == "REDACTED" || boundary.EndsWith("00000000-0000-0000-0000-000000000000"))
+            if (boundary == "REDACTED" || boundary == "BOUNDARY" || boundary.EndsWith("00000000-0000-0000-0000-000000000000"))
```
Comment (alzimmermsft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/MultipartUtilities.cs:113]: Do all other matching systems work like this where there's a hidden set of allowed redacted values?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordMatcher.cs (hunk 1 above)
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordMatcher.cs]: Possibly, but it's some nebulous "maybe." Realistically swapping the setters to private changes nothing and is more "correct".

```suggestion
        public bool ShouldCompareBodies { get; private set; }

        /// <summary>
        /// When true, query parameter ordering is ignored during URI normalization.
        /// </summary>
        public bool ShouldIgnoreQueryOrdering { get; private set; }
```

Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordMatcher.cs]: This responds to your feedback tho 👍

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/MultipartUtilities.cs (hunk 2 above)
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/MultipartUtilities.cs:113]: This is backwards compatibility with all our existing recordings. I simply don't have the time to go back and re-record all of these when I have another looming deadline.

Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/MultipartUtilities.cs:113]: The proper solution is to eliminate Content-Type: --boundary=X sanitization, as we modify the Content-Type headers to ignore those boundary values now that we properly compare body bits minus boundaries during matching.

**Review by benbp — APPROVED** [review]
